src/clims/services/workflow/__init__.py

```diff
@@ -0,0 +1,546 @@
+        batch.save()
+
+        for task in tasks:
+            task.tracked_object._archetype.work_batches.add(batch)
+            task.tracked_object._archetype.save()
+
+        # TODO(clims-345): Make sure that we're not adding a substance to another workbatch if
+        # it's already "locked".
+        return batch
+
+    def get_tasks_by_ids(self, task_ids):
+        task_ids_by_provider = defaultdict(list)
+
+        for task_id in task_ids:
+            provider_id, id = task_id.split("/")
+            task_ids_by_provider[provider_id].append(id)
+
+        tasks = list()
+        for provider_id, ids in task_ids_by_provider.items():
+            handler = self.handlers[provider_id]["handler"]
+            tasks.extend(handler.get_tasks_by_ids(ids))
+        self._materialize_tracked_objects(tasks)
+        return tasks
+
+    def get_handler(self, workflow_cls):
+        for handler_config in self.handlers.values():
+            configured_class = handler_config["cls"]
+            if issubclass(workflow_cls, configured_class):
+                return handler_config["handler"]
+            else:
+                raise NotImplementedError(
+                    "Workflow class '{}' has no handler".format(workflow_cls))
+
+    def AssignmentModel(self, item, user):
+        # Creates an "assignment" model based on which item we're assigning to a workflow
+        # This model is a historical record in the database about which assignment was made,
+        # by whom, when etc.
+        from clims.services.substance import SubstanceBase
+        from clims.models import SubstanceAssignment
+
+        # The DB model we'll hook up to:
+        model = item._archetype
+
+        if not model.id:
+            raise AssignmentError(
+                "Can't create an assignment model for unsaved model: {}".
+                format(model))
+
+        if isinstance(item, SubstanceBase):
+            return SubstanceAssignment(substance=model, user=user)
+        else:
+            raise AssignmentError("Unexpected item {}".format(item))
+
+    def assign_item(self, workflow, item, user):
+        from clims.models import SubstanceAssignment
+
+        # Fetch the handler that takes care of creating the workflow in the external workflow engine
+        handler = self.get_handler(workflow.__class__)
+        assignment = self.AssignmentModel(item, user)
+
+        # Fetch the model we use to keep a history of the item
+        # logger.info("Assigning {} to {} with vars {}".format(
+        #     item, workflow_class, {}))
+
+        # NOTE: We don't want the assignment to be saved in a transaction. The reason is that
+        # we want an uncompleted item in the assignment table to be able to identify workflows
+        # we might have started or not, of which we don't have any record. This is necessary
+        # since the workflow engine is outside of our database transactions.
+        require_no_transaction()
```

Comment:
This is basically just an assert. It will raise an error if we are in a transaction (assert_no_transaction would be clearer!)

This is a trick I'm not sure we really require, because tests could also "ensure" that this doesn't happen.

The thing is that I have three sequential actions, (POSTGRES1, CAMUNDA, POSTGRES2). I want to ensure that if POSTGRES2 fails, I still have a record of POSTGRES1. If they are in a transaction, that would not happen, because both are rollbacked.

Reply:
Btw, I changed it to assert_no_transaction.